mailbox/store/src/test/java/org/apache/james/mailbox/store/mail/model/FlagsFactoryTest.java

```diff
@@ -0,0 +1,105 @@
+ * under the License.                                           *
+ ****************************************************************/
+package org.apache.james.mailbox.store.mail.model;
+
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.assertj.core.api.Assertions.assertThatThrownBy;
+
+import javax.mail.Flags;
+import javax.mail.Flags.Flag;
+
+import org.junit.Before;
+import org.junit.Test;
+
+import com.google.common.collect.ImmutableList;
+
+public class FlagsFactoryTest {
+
+    private ImmutableList<Flag> listOfSystemFlags = ImmutableList.of(Flag.SEEN, Flag.RECENT, Flag.ANSWERED);
+    private ImmutableList<String> listOfUserFlags = ImmutableList.of("userFlag", "soCool");
+    private Flags emptyFlags;
+    private Flags someFlags;
+
+    @Before
+    public void setup() {
+        emptyFlags = new Flags();
+        someFlags = new Flags();
+        listOfSystemFlags.forEach(someFlags::add);
+        listOfUserFlags.forEach(someFlags::add);
+    }
+
+    @Test
+    public void builderShouldThrowWhenNoFlagsProvided() {
+        assertThatThrownBy(() -> FlagsFactory.builder().build()).isInstanceOf(IllegalStateException.class);
+    }
+
+    @Test
+    public void builderShouldNotRequireFlagsInstanceWhenUserFlagsDefined() {
+        Flags actual = FlagsFactory.builder().addUserFlags("userFlag").build();
+        assertThat(actual.getUserFlags()).containsOnly("userFlag");
+    }
+
+
+    @Test
+    public void builderShouldNotRequireUserFlagsWhenFlagsInstanceDefined() {
+        assertThat(FlagsFactory.builder().flags(new Flags()).build()).isNotNull();
+    }
+
+    @Test
+    public void builderShouldFilterUserFlags() {
+        Flags actual = FlagsFactory.builder()
+            .flags(someFlags)
+            .filteringFlags(
+                FlagsFilter
+                    .builder()
+                    .userFlagFilter(f -> f.equals("soCool"))
+                    .build())
+            .build();
+        assertThat(actual.getUserFlags()).containsOnly("soCool");
+    }
+
+    @Test
+    public void builderShouldFilterSystemFlags() {
+        Flags actual = FlagsFactory.builder()
+            .flags(someFlags)
+            .filteringFlags(
+                FlagsFilter
+                    .builder()
+                    .systemFlagFilter(f -> f.equals(Flag.SEEN))
+                    .build())
+            .build();
+        assertThat(actual.getSystemFlags()).containsOnly(Flag.SEEN);
+    }
+
+    @Test
+    public void builderShouldAllowFilteringOnEmptyFlags() {
+        Flags actual = FlagsFactory.builder()
+            .flags(emptyFlags)
+            .filteringFlags(
+                FlagsFilter
+                    .builder()
+                    .systemFlagFilter(f -> f.equals(Flag.SEEN))
+                    .userFlagFilter(f -> f.equals("soCool"))
+                    .build())
+            .build();
+        assertThat(actual).isNotNull();
+    }
+
```

Comment:
can you add a test where you are filtering at the same time on user flags and on system flags (same as this one, but with someflags).